### PR TITLE
Validate GPU requirements against feature gate

### LINKED ISSUE #55679: Pods with GPU requests should be rejected early if Accelerators feature gate is off
/kind bug

*What happened**:
When requesting a Pod creation with alpha.kubernetes.io/nvidia-gpu specified, I was hoping that the creation is a failure because I didn't enable the Accelerators feature gate. However, it succeeded. The pod is stuck in Pending status which is fine.

**What you expected to happen**:
I'm expecting that a Pod requesting GPU will be rejected when its container resource requirements are validated. Having the GPU usage guarded by the feature gate will help cleaning removing it in future when device plugin stabilizes.

**How to reproduce it (as minimally and precisely as possible)**:
Disable Accelerators feature gate and create a Pod with GPU request or limit, the pod is successfully created. It means that the feature gate is not doing its job.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by tengqm**
**What this PR does / why we need it**:
Check if Accelerators feature-gate is on and reject Pod creation if:

- there are GPU requests or limits, and
- the `Accelerators` feature-gate is not enabled

**Which issue(s) this PR fixes**:
Fixes #55679 

**Special notes for your reviewer**:

**Release note**:
```release-note
Pod with GPU resource requests and/or limits now fails validation if the `Accelerators` feature gate is not enabled.
```

### CODE REVIEW

**Comment by vishh** [comment]
The resource name for GPUs is changing since v1.8.0. I don't think this validation makes sense. @jiayingz perhaps we should send out an announcement that the existing `alpha.kubernetes.io/nvidia-gpu` resource is already deprecated and will be removed in v1.10?
Deprecation can also be made obvious by failing validation along with an opt-out of that validation.

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -4025,24 +4025,28 @@ func ValidateResourceRequirements(requirements *core.ResourceRequirements, fldPa
...
+			allErrs = append(allErrs, field.Forbidden(limPath, fmt.Sprintf("%s field disabled by feature-gate for ResourceRequirements", resourceName)))
+		}
 	}
 	for resourceName, quantity := range requirements.Requests {
 		fldPath := reqPath.Key(string(resourceName))
 		// Validate resource name.
 		allErrs = append(allErrs, validateContainerResourceName(string(resourceName), fldPath)...)
 		// Validate resource quantity.
 		allErrs = append(allErrs, ValidateResourceQuantityValue(string(resourceName), quantity, fldPath)...)
+		// GPU might be disable by feature-gate
```
Comment (tallclair) [inline pkg/apis/core/validation/validation.go]: Are the ephemeral storage & huge page checks missing here too? If so these should all be extracted to a helper method.

**Comment by tengqm** [comment]
Thanks for the review comments. Since GPU support is being rebased on device plugins, the `Accelerators` feature gate will be removed together. But that happens in v1.10, we need to make sure the current implementation is still consistent. I mean, the `Accelerators` gate should be gating resource GPU requests/limits.

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (tengqm) [reply pkg/apis/core/validation/validation.go]: good point. ephemeral storage should be gated by `LocalSotrageCapacityIsolation` and huge pages should be gated by `HugePages`.

**Comment by vishh** [comment]
This PR is incorrect because the new resource name is `nvidia.com/gpu` and if we do add validation it should be for that resource name. However that resource name is an extended resource and so k8s doesn't do any validation by default.

**Comment by tengqm** [comment]
Okay. I think I get the idea, maybe I am still not, ;)  Please bear with me.

GPU was named "`alpha.kubernetes.io/nvidia-gpu`" before device plugin is there, and it should be guarded by the `Accelerators=True` gate (*but not yet*).  The related code is being deprecated but not removed. People (e.g. some of our users) are still using it. This PR is trying to get the validation done right, even though the whole "alpha" thing will go away once device plugin stabilize.

In the new proposal, GPU is treated as "extended resources" implemented using device plugins. Its resource name has been changed to "nvidia.com/gpu" on GCE. This resource type is not checked because we don't have an universal naming convention. Even if it can be checked against feature gate, it should be checked against `DevicePlugins=true`, not `Accelerators=true`.  This PR is not about the new thing.

**Comment by mindprince** [comment]
/close

This is deprecated in 1.10 and going to be removed in 1.11: https://github.com/kubernetes/kubernetes/pull/61498

**Pull request closed without merging**